### PR TITLE
Supplier adjustment model

### PR DESCRIPTION
**Pull request by igorsergiichuk**
## Overview

| Q                 | A
| ----------------- | ---
| Bugfix?           | no
| New feature?      | yes
| Breaking Changes? | no
| Fixed issues      | #

## The problem

SupplierAdjustment model was missing

## How I resolved it

Added SupplierAdjustment model + tests

## How to verify it

Test it!

Notify the following people: @darrynten

### CODE REVIEW

File: src/Models/SupplierAdjustment.php [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+ * @package  SageOne
+ * @author   Ihor Sergiichuk <igorsergiichuk@gmail.com>
+ * @license  MIT <https://github.com/darrynten/sage-one-php/blob/master/LICENSE>
+ * @link     https://github.com/darrynten/sage-one-php
+ */
+
+namespace DarrynTen\SageOne\Models;
+
+use DarrynTen\SageOne\BaseModel;
+use DarrynTen\SageOne\Models\ModelCollection;
```
Comment (darrynten) [inline src/Models/SupplierAdjustment.php]: Why do you need the collection here?

File: src/Models/SupplierAdjustment.php [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+     *
+     * @var array $features
+     */
+    protected $features = [
+        'all' => true,
+        'get' => true,
+        'save' => true,
+        'delete' => false,
+    ];
+}
```
Comment (darrynten) [inline src/Models/SupplierAdjustment.php:188]: You're missing custom methods again, the Calculate call is not yet implemented here

https://accounting.sageone.co.za/api/1.1.2/Help/Api/POST-SupplierAdjustment-Calculate

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+
+namespace DarrynTen\SageOne\Tests\SageOne\Models;
+
+use DarrynTen\SageOne\Models\Supplier;
+use DarrynTen\SageOne\Models\SupplierAdjustment;
+use DarrynTen\SageOne\Models\Account;
+use DarrynTen\SageOne\Models\AccountCategory;
+use DarrynTen\SageOne\Models\TaxType;
+use DarrynTen\SageOne\Models\SupplierCategory;
+use DarrynTen\SageOne\Request\RequestHandler;
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: I don't see this import being used here?

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 4]
```diff
@@ -0,0 +1,892 @@
...
+            $this->assertEquals($model1->total, 10.0);
+            $this->assertEquals($model1->contraAccountId, 11);
+            $this->assertEquals($model1->memo, 'sample string 12');
+            $this->assertTrue($model1->hasAttachments);
+            $this->assertInstanceOf(Supplier::class, $model1->Supplier);
+            $this->assertEquals($model1->Supplier->name, 'sample string 1');
+            $this->assertInstanceOf(SupplierCategory::class, $model1->Supplier->category);
+            $this->assertEquals($model1->Supplier->category->description, 'sample string 1');
+            $this->assertEquals($model1->Supplier->category->id, 2);
+            $this->assertEquals($model1->Supplier->category->modified->format('Y-m-d'), '2017-07-18');
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: `Supplier` should be `supplier` with no capital S

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 5]
```diff
@@ -0,0 +1,892 @@
...
+                'type' => 'string',
+                'nullable' => false,
+                'readonly' => false,
+            ],
+            'hasAttachments' => [
+                'type' => 'boolean',
+                'nullable' => false,
+                'readonly' => false,
+            ],
+            'Supplier' => [
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: This mustn't have a capital S

File: tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json [hunk 6]
```diff
@@ -0,0 +1,256 @@
+{
+  "TotalResults": 1,
```
Comment (darrynten) [inline tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json]: You didn't change this number to be bigger than the returned results, see #45

File: tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+      "TaxPeriodId": 1,
+      "TotalOutstanding": 1.0,
+      "Inclusive": true,
+      "Supplier_CurrencyId": 1,
+      "Supplier_ExchangeRate": 1.0,
+      "Link_CurrencyId": 1,
+      "Link_ExchangeRate": 1.0
+    },
+    {
+      "ID": 1,
```
Comment (darrynten) [inline tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json]: You didn't change the details in this second result to be different to the first result, see #45

**Review by darrynten — CHANGES_REQUESTED** [review]
Some good work, a few changes, but is also missing the crucial Calculate function (I know @make-it-git is working on something along these lines now) so this still needs to come in here

File: src/Models/SupplierAdjustment.php (hunk 1 above)
Comment (igorsergiichuk) [reply src/Models/SupplierAdjustment.php]: deleted, redundant

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php (hunk 3 above)
Comment (igorsergiichuk) [reply tests/SageOne/Models/SupplierAdjustmentModelTest.php]: AccountCategory - take a look at line 286;
TaxType - take a look at line 268;
SupplierCategory - take a look at line 216;
RequestHandler will be deleted as redundant.

File: src/Models/SupplierAdjustment.php (hunk 2 above)
Comment (darrynten) [reply src/Models/SupplierAdjustment.php:188]: See #48 for insight into calculate, also sent details on slack

Comment (igorsergiichuk) [reply src/Models/SupplierAdjustment.php:188]: thanks. added

File: tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json (hunk 6 above)
Comment (igorsergiichuk) [reply tests/mocks/SupplierAdjustment/GET_SupplierAdjustment_Get.json]: Done!

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 8]
```diff
@@ -888,4 +888,142 @@ public function testSave()
...
+    public function testCalculate()
+    {
+        $model = $this->setUpRequestMock(
+            'POST',
+            SupplierAdjustment::class,
+            'SupplierAdjustment/Calculate',
+            'SupplierAdjustment/POST_SupplierAdjustment_Calculate_REQ.json',
+            'SupplierAdjustment/POST_SupplierAdjustment_Calculate_RESP.json'
+        );
+        $modelData= json_decode(file_get_contents(__DIR__ . '/../../mocks/SupplierAdjustment/POST_SupplierAdjustment_Calculate_REQ.json'));
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: Missing space before the `=`

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 9]
```diff
@@ -888,4 +888,142 @@ public function testSave()
...
+        $model->loadResult($modelData);
+        $calculated = $model->calculate();
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: Could you possibly add tests here for the state of the `$calculated` result directly after `calculate()` is called and before you load it into the next model?

**Review by darrynten — CHANGES_REQUESTED** [review]
Some minor changes, but nice job!

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php (hunk 9 above)
Comment (igorsergiichuk) [reply tests/SageOne/Models/SupplierAdjustmentModelTest.php]: done

File: tests/SageOne/Models/SupplierAdjustmentModelTest.php [hunk 10]
```diff
@@ -898,11 +898,137 @@ public function testCalculate()
             'SupplierAdjustment/POST_SupplierAdjustment_Calculate_REQ.json',
             'SupplierAdjustment/POST_SupplierAdjustment_Calculate_RESP.json'
         );
-        $modelData= json_decode(file_get_contents(__DIR__ . '/../../mocks/SupplierAdjustment/POST_SupplierAdjustment_Calculate_REQ.json'));
+        $modelData = json_decode(file_get_contents(__DIR__ . '/../../mocks/SupplierAdjustment/POST_SupplierAdjustment_Calculate_REQ.json'));
         $model->loadResult($modelData);
         $calculated = $model->calculate();
+
+        $this->assertEquals($model->id, 1);
```
Comment (darrynten) [inline tests/SageOne/Models/SupplierAdjustmentModelTest.php]: You need to test what `$calculated` looks like, not what `$model` looks like

**Review by darrynten — CHANGES_REQUESTED** [review]

**Review by darrynten — APPROVED** [review]
Congratulations on your first 'complex' model! 🍾

**Review by darrynten — CHANGES_REQUESTED** [review]
Wait, needs a checkmark in README

**Review by darrynten — APPROVED** [review]

**Pull request merged**